arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,504 @@
+
+  switch (in)
+    {
+      case INSN_C_LW:
+        imm = insn.lw.imm2 << 2 | insn.lw.imm53 << 3 | insn.lw.imm6 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.lw.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.lw.rs1] + imm;
+        ctx->len  = 4;
+        ctx->sext = true;
+        *(uintptr_t *)ctx->dest = 0; /* Zero the target register */
+        break;
+
+      case INSN_C_LWSP:
+        imm = insn.lwsp.imm42 << 2 | insn.lwsp.imm5 << 5 |
+              insn.lwsp.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.lwsp.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 4;
+        ctx->sext = true;
+        *(uintptr_t *)ctx->dest = 0; /* Zero the target register */
+        break;
+
+      case INSN_C_LD:
+        imm = insn.ld.imm53 << 3 | insn.ld.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.ld.rs1] + imm;
+        ctx->len  = 8;
+        *(uintptr_t *)ctx->dest = 0; /* Zero the target register */
+        break;
+
+      case INSN_C_LDSP:
+        imm = insn.ldsp.imm43 << 3 | insn.ldsp.imm5 << 5 |
+              insn.ldsp.imm86 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 8;
+        *(uintptr_t *)ctx->dest = 0; /* Zero the target register */
+        break;
+
+      case INSN_C_SW:
+        imm = insn.sw.imm2 << 2 | insn.sw.imm53 << 3 | insn.sw.imm6 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sw.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SWSP:
+        imm = insn.swsp.imm52 << 2 | insn.swsp.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.swsp.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SD:
+        imm = insn.sd.imm53 << 3 | insn.sd.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sd.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SDSP:
+        imm = insn.sdsp.imm53 << 3 | insn.sdsp.imm86 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.sdsp.rs2];
+        ctx->len  = 8;
+        break;
+
+      default:
+        _alert("Compressed: %x\n", insn.insn);
+        return false;
+    }
+
+  regs[REG_EPC] += 2;
+
+  return true;
+#else
+  return false;
+#endif
+}
+
+/****************************************************************************
+ * Name: decode_insn_ctx
+ *
+ * Description:
+ *   Try to decode a normal encoding
+ *
+ ****************************************************************************/
+
+static bool decode_insn_ctx(uintptr_t *regs, riscv_insn_ctx_t *ctx)
+{
+  uint32_t in;
+  rv_insn_t insn;
+  int32_t imm;
+
+  insn.insn = *(uint32_t *)regs[REG_EPC];
```

Comment:
should we handle the unaligned address here?

Reply:
Don't need since :
```
The C extension allows 16-bit instructions to be freely intermixed with 32-bit instructions, 
with the latter now able to start on any 16-bit boundary, i.e., IALIGN=16. 
With the addition of the C extension, no instructions can raise instruction-address-misaligned exceptions.
```
And in this patch I'm trying to handle odd address boundary memory access.

Reply:
Yes, fetch unit can handle the instruction-address-misaligned case, but read 4 bytes at line 370 is done by execution unit, not fetch unit. The execution unit shouldn't have the capability to handle the unalignment access if riscv_misaligned is enabled.

Reply:
So, let fetch Instruction by memcpy.